### PR TITLE
[NanoAOD] Add PF constituents for AK8 jets

### PR DESCRIPTION
**Pull request by nurfikri89**
#### PR description:

This PR adds PF candidate constituents for AK8 jets with abs(eta) < 2.5. Two tables are added: 

- `PFCand` : PF candidates with minimal information (`pt`, `eta`, `phi`, `mass`, `pdgId`).
- `FatJetPFCand`: Mapping information of the candidates in `PFCand` to  the AK8 jets in `FatJet`.

This new feature allows analyzers to perform Lund Plane Reweighting ([DP-2023-046](https://twiki.cern.ch/twiki/bin/view/CMSPublic/PhysicsResultsDP23046)) and calculate substructure variables for AK8 jets, like the Lund Plane itself. The feature was discussed during XPOG meeting ([21/08/24](https://indico.cern.ch/event/1440336/#9-nanoaod-modification-for-lun)), where the event size increase estimate was presented, and NanoAOD Deep Dive ([3/09/24](https://indico.cern.ch/event/1434848/#33-b2g-and-nanoaod-20-10)).

#### PR validation:

- passes the usual runTheMatrix test: `runTheMatrix.py -l limited -i all --ibeos`
- passes MiniAOD workflows: `runTheMatrix.py -i all --ibeos  -l 2500.021,2500.022,2500.023,2500.024,2500.031,2500.032,2500.033,2500.034`

### CODE REVIEW

**Comment by jfernan2** [comment]
please test

**Comment by nurfikri89** [comment]
Apologies. Typo fixed.

**Comment by hqucms** [comment]
please test

**Comment by jfernan2** [comment]
+1

File: PhysicsTools/NanoAOD/python/nanoDQM_cfi.py [hunk 1]
```diff
@@ -240,6 +240,17 @@
...
+            sels = cms.PSet(),
+            plots = cms.VPSet(
+                Count1D('_size', 10, 0, 100, 'PF candidates'),
+                Plot1D('pt', 'pt', 20,  0, 50, 'Puppi-weighted pt'),
+                Plot1D('eta', 'eta', 20, -4,  4., 'eta'),
+                Plot1D('phi', 'phi', 20, -3.14159, 3.14159, 'phi'),
+                Plot1D('mass', 'mass', 10, 0, 1., 'Puppi-weighted mass'),
+                Plot1D('pdgId', 'pdgId', 44, -220, 220, 'PF candidate type (+/-211 = ChgHad, 130 = NeuHad, 22 = Photon, +/-11 = Electron, +/-13 = Muon, 1 = HFHad, 2 = HFEM)'),
+            )
+        ),
```
Comment (hqucms) [inline PhysicsTools/NanoAOD/python/nanoDQM_cfi.py:253]: @nurfikri89 Would it make sense to add `FatJetPFCand_PFCandIdx` and `FatJetPFCand_jetIdx` to DQM as well?

Comment (nurfikri89) [reply PhysicsTools/NanoAOD/python/nanoDQM_cfi.py:253]: Looking at the indices that we store for other objects, we do not make DQM plots for them. I do not see a reason to do anything differently here.

File: PhysicsTools/NanoAOD/python/jetsAK8_cff.py [hunk 2]
```diff
@@ -317,10 +317,50 @@ def nanoAOD_addDeepInfoAK8(process, addDeepBTag, addDeepBoostedJet, addDeepDoubl
 fatJetTable.variables.pt.precision=10
 subJetTable.variables.pt.precision=10
 
+##############################################################
+# AK8 constituents
+###############################################################
+finalJetsAK8PFConstituents = cms.EDProducer("PatJetConstituentPtrSelector",
+    src = cms.InputTag("finalJetsAK8"),
+    cut = cms.string("abs(eta) <= 2.5")
```
Comment (hqucms) [inline PhysicsTools/NanoAOD/python/jetsAK8_cff.py:325]: Would it be better to use the following to ensure consistency?
```suggestion
    src = fatJetTable.src,
    cut = fatJetTable.cut
```

File: PhysicsTools/NanoAOD/python/jetsAK8_cff.py [hunk 3]
```diff
@@ -317,10 +317,50 @@ def nanoAOD_addDeepInfoAK8(process, addDeepBTag, addDeepBoostedJet, addDeepDoubl
...
+    )
+)
+
+finalJetsAK8ConstituentsTable = cms.EDProducer("SimplePatJetConstituentTableProducer",
+    name = cms.string(fatJetTable.name.value()+"PFCand"),
+    candIdxName = cms.string("PFCandIdx"),
+    candIdxDoc = cms.string("Index in the PFCand table"),
+    jets = fatJetTable.src,
+    candidates = pfCandidatesTable.src,
+    jetCut = cms.string("")
```
Comment (hqucms) [inline PhysicsTools/NanoAOD/python/jetsAK8_cff.py:356]: Would it be better to use
```suggestion
    jetCut = fatJetTable.cut
```
to ensure consistency? It does not matter now as both `finalJetsAK8` and `fatJetTable` apply `pt > 170`, but just to be on the safe side.

File: PhysicsTools/NanoAOD/python/jetsAK8_cff.py (hunk 2 above)
Comment (nurfikri89) [reply PhysicsTools/NanoAOD/python/jetsAK8_cff.py:325]: For `src`,  you are right. I will correct it but for `cut`, we explicitly set `"abs(eta) <= 2.5"` to only store the PF candidates for AK8 jets within that eta range.

File: PhysicsTools/NanoAOD/python/jetsAK8_cff.py (hunk 3 above)
Comment (nurfikri89) [reply PhysicsTools/NanoAOD/python/jetsAK8_cff.py:356]: I will correct this also.

File: PhysicsTools/NanoAOD/python/jetsAK8_cff.py (hunk 2 above)
Comment (hqucms) [reply PhysicsTools/NanoAOD/python/jetsAK8_cff.py]: Why explicitly requiring `abs(eta) <= 2.5`? IIRC in MiniAOD it's already `rapidity<2.4`, so it's largely similar but there could be some edge cases?

**Comment by hqucms** [comment]
please test

File: PhysicsTools/NanoAOD/python/jetsAK8_cff.py (hunk 2 above)
Comment (nurfikri89) [reply PhysicsTools/NanoAOD/python/jetsAK8_cff.py]: There is a selection on the jets which we retrieve the their constituents to run soft-drop grooming: 

https://github.com/cms-sw/cmssw/blob/6f9d2fd122dd1a13861bc581be8d87034489f8d9/PhysicsTools/PatAlgos/python/slimming/applySubstructure_cff.py#L36-L37

and then a `pt > 100` selection to store slimmedJetsAK8:
 
https://github.com/cms-sw/cmssw/blob/6f9d2fd122dd1a13861bc581be8d87034489f8d9/PhysicsTools/PatAlgos/python/slimming/applySubstructure_cff.py#L131

so we can still have AK8 jets in the forward region as I cannot find any explicit eta cut for AK8 puppi jets in CMSSW.

**Comment by hqucms** [comment]
+1

**Comment by mandrenguyen** [comment]
+1
